src/python/gem5/simulate/simulator.py

```diff
@@ -649,3 +789,32 @@ def save_checkpoint(self, checkpoint_dir: Path) -> None:
                                will be saved.
         """
         m5.checkpoint(str(checkpoint_dir))
+
+    def _get_simobject_tree_hash(self) -> int:
+        if self._instantiated:
+            with StringIO() as f:
+                self._root.print_ini(ini_file=f)
+                return hash(f.read())
+
+        return self.hash(self.board)
```

Comment:
This seems like it's going to be a big overhead. I see two problems
1. I think it's a bad idea to write a file and then read it back to do this. We should make the `ini` generation just be a string (or develop a better hash).
2. This is going to be high overhead no matter what, We should cache this value.

Reply:
Thank you for your comments but this is still a draft RN, I'm playing around with these changes and just Github to store my progress. I'll let you know when I arrive on a solution here and for your other points. I've not executed any of this new code yet so would not be surprised if it's buggy.